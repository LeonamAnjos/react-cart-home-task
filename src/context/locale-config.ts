export type LocaleType = "de-DE" | "en-US";

type LocaleConfig = {
  currency: string;
};

export const localeMap: Record<LocaleType, LocaleConfig> = {
  "de-DE": { currency: "EUR" },
  "en-US": { currency: "USD" },
};
