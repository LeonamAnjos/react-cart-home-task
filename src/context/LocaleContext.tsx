import React, { createContext, ReactNode, useMemo, useState } from "react";
import { localeMap, LocaleType } from "./locale-config";

type LocaleState = {
  getLocale: () => LocaleType | undefined;
  setLocale: (locale: LocaleType) => void;
  currencyFormat: (num: number) => string;
};

const LocaleContext = createContext<LocaleState>({
  getLocale: (): LocaleType | undefined => undefined,
  setLocale: (locale: LocaleType): void => {
    /* EMPTY */
  },
  currencyFormat: (num: number): string => "",
});

export interface LocaleProviderProps {
  children?: ReactNode;
  locale: LocaleType;
}

export const LocaleProvider = ({ children, locale }: LocaleProviderProps) => {
  const [value, setValue] = useState(locale);

  const contextValue = useMemo(() => {
    const formatter = new Intl.NumberFormat(value, {
      style: "currency",
      currency: localeMap[value]?.currency,
      minimumFractionDigits: 2,
    });

    return {
      getLocale: () => value,
      setLocale: (locale: LocaleType) => setValue(locale),
      currencyFormat: (num: number): string => formatter.format(num),
    };
  }, [value, setValue]);

  return (
    <LocaleContext.Provider value={contextValue}>
      {children}
    </LocaleContext.Provider>
  );
};

export default LocaleContext;
