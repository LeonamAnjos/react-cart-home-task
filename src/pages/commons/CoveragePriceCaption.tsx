import React, { useContext } from "react";
import Grid from "@mui/material/Grid";
import LocaleContext from "../../context/LocaleContext";
import Caption from "../../components/Caption";

type CoveragePriceCaptionProps = {
  coverage: number;
  price: number;
};

const CoveragePriceCaption = ({
  coverage,
  price,
}: CoveragePriceCaptionProps) => {
  const { currencyFormat } = useContext(LocaleContext);

  return (
    <Grid container spacing={2}>
      <Grid item md={6}>
        <Caption>{`Coverage: ${currencyFormat(coverage)}`}</Caption>
      </Grid>
      <Grid item md={6}>
        <Caption>{`Price: ${currencyFormat(price)}`}</Caption>
      </Grid>
    </Grid>
  );
};

export default CoveragePriceCaption;
