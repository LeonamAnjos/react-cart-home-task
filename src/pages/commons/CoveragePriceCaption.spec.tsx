import { render } from "@testing-library/react";
import { LocaleProvider } from "../../context/LocaleContext";
import CoveragePriceCaption, {
  CoveragePriceCaptionProps,
} from "./CoveragePriceCaption";

describe("<CoveragePriceCaption />", () => {
  const customRender = (props?: Partial<CoveragePriceCaptionProps>) => {
    return render(
      <LocaleProvider locale="en-US">
        <CoveragePriceCaption
          coverage={123.45}
          price={12.34}
          {...props}
        ></CoveragePriceCaption>
      </LocaleProvider>
    );
  };
  it("SHOULD render", () => {
    const { container } = customRender();

    expect(container).toMatchSnapshot();
  });
});
