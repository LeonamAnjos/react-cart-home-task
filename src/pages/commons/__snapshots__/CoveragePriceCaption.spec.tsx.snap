// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CoveragePriceCaption /> SHOULD render 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-6 css-1k7sk4d-MuiGrid-root"
    >
      <div
        class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-bjfanr-MuiTypography-root"
      >
        Coverage: $123.45
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-6 css-1k7sk4d-MuiGrid-root"
    >
      <div
        class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-bjfanr-MuiTypography-root"
      >
        Price: $12.34
      </div>
    </div>
  </div>
</div>
`;
